subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_adv_firmware.c

```diff
@@ -0,0 +1,460 @@
+			error = true;
+		}
+		break;
+	case STATE_IDLE:
+		if (update_state[obj_inst_id] == STATE_DOWNLOADING) {
+			lwm2m_engine_set_u8(path, 10);
+		}
+		break;
+	default:
+		LOG_ERR("Unhandled state: %u", state);
+		return;
+	}
+
+	if (error) {
+		LOG_ERR("Invalid state transition: %u -> %u", update_state[obj_inst_id], state);
+	}
+
+	path = lwm2m_adv_path(obj_inst_id, FIRMWARE_STATE_ID);
+	lwm2m_engine_set_u8(path, state);
+
+	LOG_DBG("Update %s state = %d", path, state);
+	now = time(NULL);
+	lwm2m_engine_set_time(lwm2m_adv_path(obj_inst_id, FIRMWARE_LAST_STATE_CHANGE_TIME_ID), now);
+}
+
+uint8_t lwm2m_adv_firmware_get_update_result(uint16_t obj_inst_id)
+{
+	return update_result[obj_inst_id];
+}
+
+void lwm2m_adv_firmware_set_update_result(uint16_t obj_inst_id, uint8_t result)
+{
+	uint8_t state;
+	bool error = false;
+	const char *path;
+
+	switch (result) {
+	case RESULT_DEFAULT:
+		/* Usually this is cancell operation */
+		if (update_state[obj_inst_id] != STATE_IDLE) {
+			result = LWM2M_ADV_FOTA_CANCELLED;
+		}
+
+		lwm2m_adv_firmware_set_update_state(obj_inst_id, STATE_IDLE);
+		break;
+	case RESULT_SUCCESS:
+		if (update_state[obj_inst_id] != STATE_UPDATING) {
+			error = true;
+			state = update_state[obj_inst_id];
+		}
+
+		lwm2m_adv_firmware_set_update_state(obj_inst_id, STATE_IDLE);
+		break;
+	case RESULT_NO_STORAGE:
+	case RESULT_OUT_OF_MEM:
+	case RESULT_CONNECTION_LOST:
+	case RESULT_UNSUP_FW:
+	case RESULT_INVALID_URI:
+	case RESULT_UNSUP_PROTO:
+		if (update_state[obj_inst_id] != STATE_DOWNLOADING) {
+			error = true;
+			state = update_state[obj_inst_id];
+		}
+
+		lwm2m_adv_firmware_set_update_state(obj_inst_id, STATE_IDLE);
+		break;
+	case RESULT_INTEGRITY_FAILED:
+		if (update_state[obj_inst_id] != STATE_DOWNLOADING &&
+		    update_state[obj_inst_id] != STATE_UPDATING) {
+			error = true;
+			state = update_state[obj_inst_id];
+		}
+
+		lwm2m_adv_firmware_set_update_state(obj_inst_id, STATE_IDLE);
+		break;
+	case RESULT_UPDATE_FAILED:
+		if (update_state[obj_inst_id] != STATE_DOWNLOADING &&
+		    update_state[obj_inst_id] != STATE_UPDATING) {
+			error = true;
+			state = update_state[obj_inst_id];
+		}
+
+		lwm2m_adv_firmware_set_update_state(obj_inst_id, STATE_IDLE);
+		break;
+	default:
+		LOG_ERR("Unhandled result: %u", result);
+		return;
+	}
+
+	if (error) {
+		LOG_ERR("Unexpected result(%u) set while state is %u", result, state);
+	}
+
+	path = lwm2m_adv_path(obj_inst_id, FIRMWARE_UPDATE_RESULT_ID);
+	lwm2m_engine_set_u8(path, result);
+
+	LOG_DBG("Update result = %d", result);
+}
+
+static int package_write_cb(uint16_t obj_inst_id, uint16_t res_id, uint16_t res_inst_id,
+			    uint8_t *data, uint16_t data_len, bool last_block, size_t total_size)
+{
+	uint8_t state;
+	int ret = 0;
+	lwm2m_engine_set_data_cb_t callback;
+
+	state = lwm2m_adv_firmware_get_update_state(obj_inst_id);
+	if (state == STATE_IDLE) {
+		/* TODO: setup timer to check download status,
```

Comment:
I assume these TODOs are remnants from the object 5.

Reply:
True